### PR TITLE
Saving books in a local file cache + util functions

### PR DESCRIPTION
**Pull request by Jatewo**
This pull request introduces a local file caching system for downloaded book texts, refactors file-saving logic for ciphers, and adds comprehensive tests for the new file utility functions. The changes improve efficiency by preventing redundant downloads, standardize file output locations, and enhance test coverage for file operations.

**Caching and file utility enhancements:**

* Added new utility functions in `utils/files.py` for saving, checking, and retrieving cached books, as well as saving ciphers to JSON files in a standardized directory structure.
* Updated `text_fetching/fetcher.py` to use the new caching utilities: before downloading a book, the code now checks if the book is already cached and loads it from disk if available; otherwise, it saves new downloads to the cache. [[1]](diffhunk://#diff-5a05d3bbb81cf5e7980a07bf7461da70813f41e76847caf6e66975496f0010f8R5) [[2]](diffhunk://#diff-5a05d3bbb81cf5e7980a07bf7461da70813f41e76847caf6e66975496f0010f8R116-R118) [[3]](diffhunk://#diff-5a05d3bbb81cf5e7980a07bf7461da70813f41e76847caf6e66975496f0010f8R142-R143)

**Refactoring and standardization:**

* Refactored `main.py` to use the new `save_cipher` utility for saving ciphers, and standardized cipher output to the `ciphers` directory. [[1]](diffhunk://#diff-2e5ad92c43aa96cc3a9cef6c6aec998b216f1379c43b1f651013d25e55989312R12) [[2]](diffhunk://#diff-2e5ad92c43aa96cc3a9cef6c6aec998b216f1379c43b1f651013d25e55989312L23-R29)
* Updated test code to expect cipher files in the `ciphers` directory and to mock file I/O where appropriate. [[1]](diffhunk://#diff-959a1effd1cb0e25016584deef9be1994af65814f318346bf3a63801b5db1636R18-R40) [[2]](diffhunk://#diff-c0469c5748e9772e6e098a1b279aceb9b03b066fb8f42cf9a818416fdc1bb607R13-R15) [[3]](diffhunk://#diff-c0469c5748e9772e6e098a1b279aceb9b03b066fb8f42cf9a818416fdc1bb607R96-R102)

**Testing improvements:**

* Added a new test suite in `test/utils/test_files.py` to cover all new file utility functions, including edge cases for missing or empty files.
* Added tests for book caching logic in the fetcher, ensuring that cached books are loaded without making HTTP requests.
* Added a test for the `Cipher` class to verify that the difficulty parameter is set correctly.

### CODE REVIEW

**Review by Fantaskink — APPROVED** [review]
LGTM

File: src/utils/files.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
+from encipherment.cipher import Cipher
+
+def save_book(book_id: str, book_text: str) -> None:
```
Comment (Mafusn) [inline src/utils/files.py:3]: Isn't it quite crazy to pass an entire book text as a parameter, maybe it should just fetch the book by id

File: test/text-fetching/test_fetcher.py [hunk 2]
```diff
@@ -90,6 +93,13 @@ def test_fetch_book_success(mocker):
         "formats": {"text/plain; charset=utf-8": "http://example.com/book.txt"}
     }
 
+    mocker.patch("text_fetching.fetcher.save_book")  # Mock save_book to avoid file I/O
```
Comment (Mafusn) [inline test/text-fetching/test_fetcher.py]: Isn't this already mocked in mock_save_book?

File: test/utils/test_files.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	from encipherment.cipher import Cipher
+	return Cipher("thisisatestcipher", 12)
+
+@pytest.fixture(autouse=True)
+def run_around_tests(tmp_path, monkeypatch):
+	monkeypatch.chdir(tmp_path) # Change to the temporary directory for the test
+	yield # Cleanup: Nothing needed as tmp_path is automatically cleaned up
+
+def test_save_book(tmp_path):
+	from utils.files import save_book
```
Comment (Mafusn) [inline test/utils/test_files.py]: Wouldn't it be cleaner to import all these functions in the top instead of in each test function?

**Review by Mafusn — CHANGES_REQUESTED** [review]
These questions likely don't need any changes.

File: test/text-fetching/test_fetcher.py (hunk 2 above)
Comment (Jatewo) [reply test/text-fetching/test_fetcher.py]: Yes, it is 👍

File: test/utils/test_files.py (hunk 3 above)
Comment (Jatewo) [reply test/utils/test_files.py]: Yea, you're probably right

File: src/utils/files.py (hunk 1 above)
Comment (Jatewo) [reply src/utils/files.py:3]: No

File: test/utils/test_files.py (hunk 3 above)
Comment (TheMOHD3886) [reply test/utils/test_files.py]: I agree with mafuzz

**Review by TheMOHD3886 — APPROVED** [review]
